src/grpc/mod.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::*;
+use futures::channel::oneshot;
+use futures::executor::block_on;
+use futures::prelude::*;
+use grpcio::{Environment, ServerBuilder};
+use std::io::Read;
+use std::net::IpAddr;
+use std::str;
+use std::sync::Arc;
+use std::{io, thread};
+
+pub mod protocol;
+
+use protocol::keyprovider;
+use protocol::keyprovider_grpc::*;
+use protocol::keyprovider_structs::*;
+
+use super::kbc_runtime;
+
+#[derive(Clone)]
+pub struct KeyProvider;
+
+impl KeyProviderService for KeyProvider {
+    fn un_wrap_key(
+        &mut self,
+        ctx: ::grpcio::RpcContext,
+        req: keyprovider::keyProviderKeyWrapProtocolInput,
+        sink: ::grpcio::UnarySink<keyprovider::keyProviderKeyWrapProtocolOutput>,
+    ) {
+        // Deserialize and parse the gRPC input to get KBC name, KBS uri and annotation.
+        let input_payload: InputPayload = match parse_input(req.KeyProviderKeyWrapProtocolInput) {
+            Ok(i) => i,
+            Err(e) => {
+                println!("Some error happend when parse the input: {}", e);
+                return ();
+            }
+        };
```

Comment:
Having traits returning a `Result` would allow for a more idiomatic, `map_err` based handling here and below.